0000-ilp-over-http.md

```diff
@@ -0,0 +1,57 @@
+---
+title: ILP Over HTTP
+draft: 1
+---
+
+# ILP Over HTTP
+> A bilateral communication protocol for server-to-server connections
+
+## Motivation
+
+Scaling Interledger infrastructure to handle large volumes of ILP packets requires horizontally scaling connectors. Using HTTP for bilateral communication enables service providers to leverage standard tools and services for hosting, load balancing, Distributed Denial of Service (DDoS) protection, and monitoring.
+
+## Overview
+
+In an ILP Over HTTP connection, both peers run HTTP servers with accessible HTTPS endpoints. When peering, the peers exchange their respective URLs, authentication tokens or TLS certificates, ILP addresses, and settlement-related details.
+
+Each ILP Prepare packet is sent as the body of an HTTP request to the peer's server endpoint. The HTTP response body will contain either the ILP Fulfill or Reject packet.
+
+## Specification
+
+This is a minimal protocol built on HTTP. HTTP/2 is RECOMMENDED for performance reasons, although HTTP/1.1 MAY also be used. Implementations SHOULD support HTTP version negotiation via Application Protocol Negotiation (ALPN).
+
+### Authentication
+
+Peers MAY use any standard HTTP authentication mechanism to authenticate incoming requests. Bearer Tokens or TLS Client Certificates are RECOMMENDED.
+
+### Request
+
+```http
+POST /ilp HTTP/2.0
```

Comment:
Is there a canonical way of writing out HTTP2 requests? It's a binary format so maybe not...

Reply:
I think that we should specify this protocol without saying a specific HTTP version. Even if HTTP2 is used on the outgoing side, it's likely that the HTTP2 connection will be made to a load balancer and then a regular HTTP connection will be made to the application that actually implements this protocol. A client might also use a proxy like envoy (https://www.envoyproxy.io/) to upgrade outgoing requests to HTTP2 while using an ordinary HTTP client themselves.

Reply:
I'll change the example, but I think it's important to mention what versions implementations SHOULD support or default to. If you're using this internally, all bets are off, but the purpose of the spec is to make sure that messages that cross administrative boundaries work.

Reply:
As discussed on the community group call, using HTTP pre-2 introduces HOL blocking that makes this likely to be significantly worse than BTP. 

I don't think using this protocol is worth it unless you use HTTP2.

Reply:
Quite possibly, and I think in most cases the libraries people would use will support HTTP2. Do you think the language around its use should be stronger?

Reply:
> As discussed on the community group call, using HTTP pre-2 introduces HOL blocking that makes this likely to be significantly worse than BTP.

HTTP with keepalive has head of line blocking, but you can just put each request on a different connection. This adds round trips, especially if TLS is used, which is why I would only imagine it inside of a datacenter/cluster or even between processes on a single machine. But I don't want to deal with the headache of certs and HTTP2 support everywhere inside of our cluster when we could just have the load balancer deal with HTTP2 and do HTTP/1.1 inside.

Reply:
> put each request on a different connection

Connection establishment is the biggest culprit in most network perf issues. The TCP connection establishment handshake is one of the main drivers behind QUIC.

> I don't want to deal with the headache of certs and HTTP2 support everywhere inside of our cluster when we could just have the load balancer deal with HTTP2 and do HTTP/1.1 inside.

What you do internally is up to you. It definitely shouldn't drive what's in the standard

Reply:
The current draft says:

```
HTTP/2 is HIGHLY RECOMMENDED for performance reasons, although HTTP/1.1 MAY also be used. Implementations SHOULD support HTTP version negotiation via Application Protocol Negotiation (ALPN).
```

@adrianhopebailie Is that okay with you?